### PR TITLE
Require the Config.swift file to be in the Tuist directory

### PR DESCRIPTION
**Pull request by mollyIV**
Resolves https://github.com/tuist/tuist/issues/693

### Short description 📝

This pull request is introducing two improvements related to a `Config` file:

* if `Config` file is found at the directory different than `Tuist`, we kindly print a warning with the suggested location of the file;
* `init` command now generates default `Config` file at `Tuist` directory.

### Implementation 👩‍💻👨‍💻

The `path` property has been added to `Config` model. That change allowed to do the linting of a path of the config file via `EnvironmentLilnter`. 

The default templates have been modified by including the default `Config` file at the `Tuist` directory. 

### Testing 

Unit tests have been added to `EnvironmentLinter`.

I did a couple of manual testing too:

```bash 
// Project without a Config file

Generating workspace MyApp.xcworkspace
Generating project MyApp
Project generated.
Total time taken: 3.911s

// Project with a Config file, but not located at Tuist directory

`Config.swift` manifest file is not located at `Tuist` directory
Generating workspace MyApp.xcworkspace
Generating project MyApp
Project generated.
Total time taken: 0.548s

// Config file located at Tuist directory

Generating workspace MyApp.xcworkspace
Generating project MyApp
Project generated.
Total time taken: 0.471s

// Config file located at Tuist directory, but invalid environment

The project, which only supports the versions of Xcode 11.0, is not compatible with your selected version of Xcode, 12.0.1
Fatal linting issues found
```

### CODE REVIEW

File: Sources/TuistGenerator/Generator/Generator.swift [hunk 1]
```diff
@@ -132,7 +132,7 @@ public class Generator: Generating {
 
     public func generateProject(at path: AbsolutePath) throws -> (AbsolutePath, Graph) {
         let config = try graphLoader.loadConfig(path: path)
-        try environmentLinter.lint(config: config).printAndThrowIfNeeded()
+        try environmentLinter.lint(config: config, at: path).printAndThrowIfNeeded()
```
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/Generator.swift:135]: @mollyIV have you considered adding a `path: AbsolutePath` attribute to the `Config` model where we set the path where the config has been loaded from. That way we don't have to modify the signature of the methods to pass down the path.

**Review by pepibumur — APPROVED** [review]
Besides my suggestion, changes look good. The warning will nudge users to move the Config to the right directory. Since this is something we'll want to deprecate for Tuist version 2.0, I wonder whether we should have a `DEPRECATIONS.md` file at the root, where we list all the functionalities that we should deprecate for the next major version. It can work as a checklist that we can go through before releasing 2.0.  What do you think @mollyIV ?

**Comment by pepibumur** [comment]
@all-contributors add @mollyIV for ideas

File: Sources/TuistGenerator/Generator/Generator.swift (hunk 1 above)
Comment (mollyIV) [reply Sources/TuistGenerator/Generator/Generator.swift:135]: 👋 Pedro, 

actually, I did. I though that `Config` model should contain only the properties that a user can define 🤔 If that's not a case, I will do the suggested change:

```
// Config.swift

public init(compatibleXcodeVersions: CompatibleXcodeVersions = .all,
                cloud: Cloud? = nil,
                generationOptions: [GenerationOptions],
                path: Path) { ... }
```

Thanks!

**Comment by mollyIV** [comment]
> Besides my suggestion, changes look good.

Thanks a lot for a review 🙇 One of the remaining point from the issue description is:

> Update the init command to generate the TuistConfig.swift file under Tuist/.

That will come in the next commits 💪

File: Sources/TuistGenerator/Generator/Generator.swift (hunk 1 above)
Comment (pepibumur) [reply Sources/TuistGenerator/Generator/Generator.swift:135]: 👋 In fact, that's the case for the `Config.swift` model in the `ProjectDescription` framework. That's the user-facing one. There's another `Config.swift` file inside the `TuistCore` framework that is an internal representation. That one can contain extra attributes, like the path to the directory where the config has been loaded from. We do exactly the same for `Project.swift` and `Workspace.swift`.

**Comment by pepibumur** [comment]
Awesome! Thanks a lot @mollyIV for tackling this :)

File: Sources/TuistGenerator/Linter/EnvironmentLinter.swift [hunk 2]
```diff
@@ -47,4 +53,21 @@ public class EnvironmentLinter: EnvironmentLinting {
...
+        guard
+            let configPath = config.path,
+            let rootDirectoryPath = rootDirectoryLocator.locate(from: configPath)
+        else {
+            return []
+        }
+
+        let tuistDirectoryPath = rootDirectoryPath.appending(RelativePath("\(Constants.tuistDirectoryName)"))
+        guard configPath.removingLastComponent() == tuistDirectoryPath else {
+            let message = "`Config.swift` manifest file is not located at `Tuist` directory"
```
Comment (mollyIV) [inline Sources/TuistGenerator/Linter/EnvironmentLinter.swift:67]: Feel free to suggest better error message 😊

File: Sources/TuistGenerator/Generator/Generator.swift (hunk 1 above)
Comment (mollyIV) [reply Sources/TuistGenerator/Generator/Generator.swift]: I implemented the suggested changes 🙇

**Comment by mollyIV** [comment]
> Since this is something we'll want to deprecate for Tuist version 2.0, I wonder whether we should have a DEPRECATIONS.md file at the root, where we list all the functionalities that we should deprecate for the next major version.

A simple checklist like that would do the job 👍 While marking the methods / pieces as deprecated, it would be nice to provide a clear guidelines for a user how to migrate though.

**Comment by mollyIV** [comment]
Oh, sorry, I am fixing the failing checks 🐇

**Review by pepibumur — APPROVED** [review]

**Comment by pepibumur** [comment]
Brilliant job @mollyIV 👏

**Pull request merged**